### PR TITLE
Add prefix lm option to MosaicGPT

### PR DESCRIPTION
**Pull request by alextrott16**
This PR introduces support for a Prefix-LM version of MosaicGPT. 
It slightly refactors the triton FA implementation to add `TritonFlashPrefixAttention` and adds logic to handle a prefix-lm-specific input `prefix_mask` (which controls which portions of the token sequence use bidirectional (inside the prefix) or causal attention).
I'm also slightly refactoring some of the code to re-use a utility for inferring the value of the required `bidirectional_mask` batch input.

### CODE REVIEW

File: examples/llm/src/models/layers/attention.py [hunk 1]
```diff
@@ -82,11 +82,17 @@ def generate_attn_mask(
...
         key_padding_mask=None,
         alibi=False,
         dtype=None,
     ):
 
+        if prefix_mask is not None:
+            raise NotImplementedError(
+                'This attention implementation does not support use of `prefix_lm`.'
+            )
+
```
Comment (vchiley) [inline examples/llm/src/models/layers/attention.py:95]: Is there anything that stops us from supporting prefix_mask for torch impl (non-causal)?

Comment (alextrott16) [reply examples/llm/src/models/layers/attention.py:95]: My laziness. I suppose I should add support for that.

Comment (alextrott16) [reply examples/llm/src/models/layers/attention.py:95]: @vchiley Actually, is it worth messing around with the torch implementation? It has a deprecation warning and is already missing support for a number of features we seem to be moving towards.

Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:95]: Given the attn rewrite, all attn impls are merged into one cls.
The only version that cannot support bias is flash and an error gets raised if you pass an attn_bias to it.

**Comment by vchiley** [comment]
converting to draft.
@alextrott16 has a new version incoming.

**Comment by abhi-mosaic** [comment]
Closing in favor of #248

**Pull request closed without merging**